### PR TITLE
Modern C++ dtor declarations

### PR DESCRIPTION
**Pull request by seanm**

### CODE REVIEW

**Comment by seanm** [comment]
I'll need someone's help to actually implement the copy assignment operators correctly.

**Review by byrnHDF — APPROVED** [review]

**Comment by seanm** [comment]
@byrnHDF would you know how to implement the copy assignment operators in the second commit?  I figure they should probably call close() like the dtor does...

**Comment by byrnHDF** [comment]
> @byrnHDF would you know how to implement the copy assignment operators in the second commit? I figure they should probably call close() like the dtor does...

I'm not up-to-date on the latest conventions for these types of functions.

**Comment by seanm** [comment]
@byrnHDF who's the HDF5 C++ expert that I can ping?

**Comment by seanm** [comment]
@gnuoyd @derobins do you know an HDF5 C++ expert that I can ping?

**Comment by bmribler** [comment]
Not calling myself an expert, but I'm reviewing this PR.

Binh-Minh
________________________________

**Comment by derobins** [comment]
> @gnuoyd @derobins do you know an HDF5 C++ expert that I can ping?

I can help, but I'm a bit underwater for the immediate future. I'll jump in as soon as I free up.

**Review by qkoziol — COMMENTED** [review]
Looks sensible, but are there tests that exercise the changes?

**Comment by seanm** [comment]
> Looks sensible, but are there tests that exercise the changes?

I didn't add any new tests, if that's what you're asking.  I imagine the dtors are well-exercised by existing tests.  As for the copy assignment operators, I'm not sure.

File: c++/src/H5AbstractDs.h [hunk 1]
```diff
@@ -68,7 +68,7 @@ class H5_DLLCPP AbstractDs {
     virtual H5std_string fromClass() const = 0;
 
     // Destructor
-    virtual ~AbstractDs();
+    virtual ~AbstractDs() = default;
```
Comment (bradenmcd) [inline c++/src/H5AbstractDs.h:70]: For a `virtual` destructor, prefer putting this in the `.cpp` file; i.e.,:
```c++
AbstractDs::AbstractDs() = default;
```
That way, you don't end up generating a definition in every translation unit that includes the header.

File: c++/src/H5ArrayType.h [hunk 2]
```diff
@@ -61,7 +61,7 @@ class H5_DLLCPP ArrayType : public DataType {
     ArrayType(const hid_t existing_id);
 
     // Noop destructor
-    virtual ~ArrayType() override;
+    virtual ~ArrayType() override = default;
```
Comment (bradenmcd) [inline c++/src/H5ArrayType.h:63]: Benign, but if you're using `override`, `virtual` is redundant.

File: c++/src/H5Exception.h [hunk 3]
```diff
@@ -75,7 +75,7 @@ class H5_DLLCPP Exception {
     Exception(const Exception &orig);
 
     // virtual Destructor
-    virtual ~Exception() throw();
+    virtual ~Exception() noexcept = default;
```
Comment (bradenmcd) [inline c++/src/H5Exception.h]: As of C++11, destructors are implicitly `noexcept` _unless the generated trivial destructor would not be_. IOW, unless a base or member has a destructor that is `noexcept(false)`.

File: c++/src/H5ArrayType.h (hunk 2 above)
Comment (seanm) [reply c++/src/H5ArrayType.h:63]: Yes, but it was already like that before my change. This PR is really just trying to fix warnings, especially in header files.

File: c++/src/H5Exception.h (hunk 3 above)
Comment (seanm) [reply c++/src/H5Exception.h]: OK, so you're saying I should just remove the `noexcept` keyword here?

File: c++/src/H5ArrayType.h (hunk 2 above)
Comment (seanm) [reply c++/src/H5ArrayType.h:63]: Searching for `virtual.*override` gives 80 hits. A cleanup of that redundancy can be for another patch.

File: c++/src/H5AbstractDs.h (hunk 1 above)
Comment (seanm) [reply c++/src/H5AbstractDs.h:70]: I have never seen a codebase do that... I can't find much discussion about it either, except for [this](https://stackoverflow.com/questions/58774272/virtual-default-destructor-definition-in-source-file-vs-in-header-file), where the answers are not so conclusive to me...

I'd like to RTFM on this if you could point me somewhere...

File: c++/src/H5Exception.h (hunk 3 above)
Comment (seanm) [reply c++/src/H5Exception.h]: Upon further research and investigation, I agree, and have made this change.

**Review by bmribler — APPROVED** [review]

**Comment by seanm** [comment]
@bmribler not sure why you just tagged this as "approved"... it's a WIP and in fact it appears to be conflicted now...

**Comment by seanm** [comment]
OK, I removed the commit that was WIP, leaving only the 1 commit that modernizes the dtors.

This is now ready for review & merge IMHO.

**Comment by lrknox** [comment]
The PR didn't run tests directly, but the commit earlier today passed all tests except the clang_format check which failed due to 2 extra blank lines, which were removed in the subsequent commit.

**Comment by bmribler** [comment]
> @bmribler not sure why you just tagged this as "approved"... it's a WIP and in fact it appears to be conflicted now...

My mistake, @seanm

**Comment by bmribler** [comment]
> I'll need someone's help to actually implement the copy assignment operators correctly.

@seanm, I'm sorry, I can do it after this modernizes the dtors PR is merged.

**Comment by seanm** [comment]
@bmribler @byrnHDF friendly ping...

**Comment by bmribler** [comment]
> @bmribler @byrnHDF friendly ping...



> @bmribler @byrnHDF friendly ping...

@seanm for the copy assignment operators?

**Comment by bmribler** [comment]
@seanm you meant for the copy assignment operators?

**Comment by seanm** [comment]
> @seanm you meant for the copy assignment operators?

No, those were moved to https://github.com/HDFGroup/hdf5/pull/3306

This PR is ready to merge IMHO.

**Review by bmribler — APPROVED** [review]

**Pull request merged**